cajita/src/Cajita_SparseIndexSpace.hpp

```diff
@@ -0,0 +1,499 @@
+    static constexpr int Rank = N;
+    //! Number of cells inside each tile, tile size reset to power of 2
+    static constexpr unsigned long long TileBitsPerDim =
+        bit_count( TileNPerDim );
+    static constexpr unsigned long long TileSizePerDim = 1 << TileBitsPerDim;
+    static constexpr unsigned long long TileMaskPerDim =
+        ( 1 << TileBitsPerDim ) - 1;
+    static constexpr unsigned long long TileBits =
+        TileBitsPerDim + TileBitsPerDim + TileBitsPerDim;
+    static constexpr unsigned long long TileSize =
+        TileSizePerDim * TileSizePerDim * TileSizePerDim;
+    //! Types
+    using KeyType = Key;     // tile hash key type
+    using ValueType = Value; // tile value type
+    using HashType = Hash;   // hash table type
+
+    // size should be global
+    SparseIndexSpace( const std::array<int, N> &size, int capacity,
+                      float rehash_factor )
+        : _blkIdSpace( 1 << bit_count( capacity ), rehash_factor, size )
+        , _tileIdSpace()
+    {
+        std::fill( _min.data(), _min.data() + Rank, 0 );
+        std::copy( size.begin(), size.end(), _max.data() );
+    }
+
+    //! given a cell ijk, insert the tile such cell reside in
+    KOKKOS_FORCEINLINE_FUNCTION
+    ValueType insert_cell( int cellid_x, int cellid_y, int cellid_z )
+    {
+        return _blkIdSpace.insert( cellid_x >> TileBitsPerDim,
+                                   cellid_y >> TileBitsPerDim,
+                                   cellid_z >> TileBitsPerDim );
+    }
+
+    KOKKOS_FORCEINLINE_FUNCTION
+    ValueType insert_tile( int tileid_x, int tileid_y, int tileid_z )
+    {
+        return _blkIdSpace.insert( tileid_x, tileid_y, tileid_z );
+    }
+
+    KOKKOS_FORCEINLINE_FUNCTION
+    ValueType query_cell( int cellid_x, int cellid_y, int cellid_z )
+    {
+        auto tileid = _blkIdSpace.find( cellid_x >> TileBitsPerDim,
+                                        cellid_y >> TileBitsPerDim,
+                                        cellid_z >> TileBitsPerDim );
+        auto cellid = _tileIdSpace.coord_to_offset(
+            cellid_x & TileMask, cellid_y & TileMask, cellid_z & TileMask );
+        return ( ValueType )( ( tileid << TileBits ) & cellid );
+    }
+
+  private:
+    //! block index space, map tile ijk to tile No
+    BlockIndexSpace<TileBitsPerDim, TileSizePerDim, TSize, HashType, Device,
+                    KeyType, ValueType>
+        _blkIdSpace;
+    //! tile index space, map cell ijk to cell local No inside a tile
+    TileIndexSpace<TileBitsPerDim, TileSizePerDim, TSize> _tileIdSpace;
+    //! space size (global), channel size
+    Kokkos::Array<int, Rank> _min;
+    Kokkos::Array<int, Rank> _max;
+};
+
+//---------------------------------------------------------------------------//
+template <unsigned long long TBits, unsigned long long TSizePerDim,
+          unsigned long long TSize, typename Hash, typename Device,
+          typename Key, typename Value>
```

Comment:
Make the template names and order consistent across the different index space types.